cardano-api/src/Cardano/Api/MetaData.hs

```diff
@@ -27,6 +30,80 @@ import qualified Data.Vector as Vector
 
 import           Shelley.Spec.Ledger.MetaData (MetaData (..), MetaDatum (..))
 
+
+
+-- | A transaction metadata validation error.
+data TxMetadataValidationError
+  = TxMetadataTextStringInvalidLengthError
+    -- ^ The length of a text string metadatum value exceeds the maximum.
+      !Int
+      -- ^ Maximum byte length.
+      !Int
+      -- ^ Actual byte length.
+  | TxMetadataByteStringInvalidLengthError
+    -- ^ The length of a byte string metadatum value exceeds the maximum.
+      !Int
+      -- ^ Maximum byte length.
+      !Int
+      -- ^ Actual byte length.
+  deriving (Eq, Show)
+
+-- | The maximum byte length of a transaction metadata text string value.
+txMetadataTextStringMaxByteLength :: Int
+txMetadataTextStringMaxByteLength = 64
+
+-- | The maximum length of a transaction metadata byte string value.
+txMetadataByteStringMaxLength :: Int
+txMetadataByteStringMaxLength = 64
+
+-- | Render a transaction metadata validation error as text.
+renderTxMetadataValidationError :: TxMetadataValidationError -> Text
+renderTxMetadataValidationError err =
+  case err of
+    TxMetadataTextStringInvalidLengthError maxLen actualLen ->
+      "Text string metadatum value must consist of at most "
+        <> show maxLen
+        <> " bytes, but it consists of "
+        <> show actualLen
+        <> " bytes."
+    TxMetadataByteStringInvalidLengthError maxLen actualLen ->
+      "Byte string metadatum value must consist of at most "
+        <> show maxLen
+        <> " bytes, but it consists of "
+        <> show actualLen
+        <> " bytes."
+
+-- | Validate the provided transaction metadata.
+validateTxMetadata
+  :: TxMetadata
+  -> Either TxMetadataValidationError TxMetadata
+validateTxMetadata txMd@(TxMetadata (MetaData mdMap)) =
+    txMd <$ for_ mdMap validate
+  where
+    validate :: MetaDatum -> Either TxMetadataValidationError ()
+    validate metaDatum =
+      case metaDatum of
+        Map mdPairs -> for_ mdPairs $ \(k, v) -> validate k >> validate v
+        List mds -> for_ mds validate
```

Comment:
Since we're playing traversable golf, can I suggest an even simpler approach:

We do not need to use monadic traversals here. We can (and should) collect all the errors. So we want to use the list _monoid_. Using a monoid lets use use monoidal `fold` from the `Foldable` class.
```
foldMap :: Monoid m => (a -> m) -> t a -> m
```

Then finally at the top level we can see if our list of errors is empty or not, and return Either a list of failures or the successful result.